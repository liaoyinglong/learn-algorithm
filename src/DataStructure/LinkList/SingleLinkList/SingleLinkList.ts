/**
 * 单列表
 */
export class SingleLinkList<T> {
  private len = 0;
  private head: Node<T> | null = null;

  get length() {
    return this.len;
  }

  add(item: T) {
    return this.inset(item, this.len);
  }

  get(targetIndex: number) {
    if (targetIndex < 0 || targetIndex > this.len) {
      return;
    }
    let i = 0;
    let node = this.head!;
    do {
      if (i === targetIndex) {
        return node;
      }
      i++;
      node = node.next!;
    } while (node);
  }

  indexOf(item: T) {
    let node = this.head;
    let i = 0;

    while (node) {
      if (node.data === item) {
        return i;
      }
      node = node.next;
      i++;
    }
    return -1;
  }

  has(item: T) {
    return this.indexOf(item) !== -1;
  }

  /**
   * @param item 待插入的元素
   * @param targetIndex 插入后的index
   * @description 需要对target-1位置上的元素进行操作
   */
  inset(item: T, targetIndex: number) {
    const pendingNode = new Node(item);
    if (targetIndex === 0) {
      pendingNode.next = this.head;
      this.head = pendingNode;
      this.len++;
      return true;
    }

    const t = this.get(targetIndex - 1);
    if (t) {
      pendingNode.next = t.next;
      t.next = pendingNode;
      this.len++;
      return true;
    }
    return false;
  }
  /**
   * @param targetIndex pending to remove index
   * @description 需要对target-1位置上的元素进行操作
   */
  remove(targetIndex: number) {
    if (targetIndex === 0) {
      this.head = this.head?.next ?? null;
      this.len--;
      return true;
    }

    const t = this.get(targetIndex - 1);
    if (t) {
      t.next = t.next?.next ?? null;
      this.len--;
      return true;
    }
    return false;
  }
}

class Node<T> {
  next: null | Node<T> = null;

  constructor(public data: T) {}
}
