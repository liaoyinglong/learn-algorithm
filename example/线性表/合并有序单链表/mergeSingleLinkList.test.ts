import { mergeSingleLinkList } from "./mergeSingleLinkList";
import { SingleLinkList } from "src/DataStructure/LinkList/SingleLinkList/SingleLinkList";

describe("合并有序顺序表", () => {
  it("合并成功应该返回l1", () => {
    const l1 = SingleLinkList.create([1, 3, 5]);
    const l2 = SingleLinkList.create([2, 4, 6]);

    expect(mergeSingleLinkList(l1, l2)).toMatchSnapshot();
  });

  it("合并成功应该返回l2", () => {
    const l1 = SingleLinkList.create([2, 4, 6]);
    const l2 = SingleLinkList.create([1, 3, 5]);

    expect(mergeSingleLinkList(l1, l2)).toMatchSnapshot();
  });

  it("其中有一个为空", () => {
    const l1 = SingleLinkList.create([]);
    const l2 = SingleLinkList.create([1, 3, 5]);

    expect(mergeSingleLinkList(l1, l2)).toMatchSnapshot();
  });
});
