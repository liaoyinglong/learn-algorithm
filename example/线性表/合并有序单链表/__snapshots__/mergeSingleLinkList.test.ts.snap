// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`合并有序顺序表 其中有一个为空 1`] = `
SingleLinkList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 3,
      "next": Node {
        "data": 5,
        "next": null,
      },
    },
  },
  "len": 3,
}
`;

exports[`合并有序顺序表 合并成功应该返回l1 1`] = `
SingleLinkList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": Node {
          "data": 4,
          "next": Node {
            "data": 5,
            "next": Node {
              "data": 6,
              "next": null,
            },
          },
        },
      },
    },
  },
  "len": 3,
}
`;

exports[`合并有序顺序表 合并成功应该返回l2 1`] = `
SingleLinkList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": Node {
          "data": 4,
          "next": Node {
            "data": 5,
            "next": Node {
              "data": 6,
              "next": null,
            },
          },
        },
      },
    },
  },
  "len": 3,
}
`;
